### PR TITLE
fix coherence error for very large tuples™

### LINKED ISSUE #131969: Diesel new coherence failure: conflicting implementations of trait
### Code

I tried this code:

```sh
git clone https://github.com/diesel-rs/diesel/
cd diesel
git checkout b170af753cc3c6774f578aae1081711a09dfa2ac
cargo +nightly check -p diesel -F postgres -F 64-column-tables
```
I expected to see this happen: Code compiles fine

Instead, this happened: Code produces the [following errors](https://gist.github.com/weiznich/98c8624e7a978e6beeea7e8f99ef635c) (Really long error, so I put it into a gist)

### Version it worked on

It most recently worked on: Beta channel (1.83)

### Version with regression

`rustc --version --verbose`:
```
rustc 1.84.0-nightly (da935398d 2024-10-19)
binary: rustc
commit-hash: da935398d582344c5b7689bd6632d8ec01b0c988
commit-date: 2024-10-19
host: x86_64-unknown-linux-gnu
release: 1.84.0-nightly
LLVM version: 19.1.1

```

Rustc-bisect points to a0c2aba29aa9ea50a7c45c3391dd446f856bef7b, so cc @lcnr  @compiler-errors  (I'm sorry to come with the next issue there)

@rustbot modify labels: +regression-from-beta-to-nightly -regression-untriaged

### PR DESCRIPTION
**Pull request by lcnr**
see https://rust-lang.zulipchat.com/#narrow/channel/364551-t-types.2Ftrait-system-refactor/topic/diesel.20error for an in-depth explanation of this issue. We once again specialize `NormalizesTo` goals to avoid the impact of erasing their expected term.

fixes #131969

r? @compiler-errors

### CODE REVIEW

File: tests/ui/traits/next-solver/overflow/coherence-alias-hang.rs [hunk 1]
```diff
@@ -1,8 +1,5 @@
 //@ check-pass
-//@ revisions: ai_current ai_next ia_current ia_next ii_current ii_next
-//@[ai_next] compile-flags: -Znext-solver
-//@[ia_next] compile-flags: -Znext-solver
-//@[ii_next] compile-flags: -Znext-solver
+//@ revisions: ai ia ii
```
Comment (compiler-errors) [inline tests/ui/traits/next-solver/overflow/coherence-alias-hang.rs:2]: Not checking w/ solver globally?

File: tests/ui/traits/next-solver/normalize/normalize-allow-too-many-vars.rs [hunk 2]
```diff
@@ -0,0 +1,52 @@
+//@ check-pass
```
Comment (compiler-errors) [inline tests/ui/traits/next-solver/normalize/normalize-allow-too-many-vars.rs:1]: Same here, not checking w/ solver globally?

File: tests/ui/traits/next-solver/overflow/coherence-alias-hang.rs (hunk 1 above)
Comment (compiler-errors) [reply tests/ui/traits/next-solver/overflow/coherence-alias-hang.rs:2]: Coherence test only

Comment (lcnr) [reply tests/ui/traits/next-solver/overflow/coherence-alias-hang.rs:2]: only testing coherence behavior

File: compiler/rustc_next_trait_solver/src/solve/eval_ctxt/canonical.rs [hunk 3]
```diff
@@ -165,12 +165,21 @@ where
...
-            }));
+        // during generalization. This prevents hangs caused by an exponential blowup,
+        // see tests/ui/traits/next-solver/coherence-alias-hang.rs.
+        //
+        // We don't do so for `NormalizesTo` goals as we erased the expected term and
+        // bailing with overflow here would prevent us from detecting a type-mismatch,
+        // causing a coherence error in diesel, see #131969. We still bail with verflow
+        // when later returning from the parent AliasRelate goal.
+        if !self.is_normalizes_to_goal {
+            let num_non_region_vars = canonical.variables.iter().filter(|c| !c.is_region()).count();
```
Comment (compiler-errors) [inline compiler/rustc_next_trait_solver/src/solve/eval_ctxt/canonical.rs:176]:
```suggestion
            let num_non_region_vars = canonical.variables.iter().filter(|c| !c.is_region() && !c.is_existential()).count();
```

Comment (compiler-errors) [reply compiler/rustc_next_trait_solver/src/solve/eval_ctxt/canonical.rs:176]: army after this

File: compiler/rustc_next_trait_solver/src/solve/eval_ctxt/canonical.rs [hunk 4]
```diff
@@ -165,12 +165,22 @@ where
...
-        if num_non_region_vars > self.cx().recursion_limit() {
-            return Ok(self.make_ambiguous_response_no_constraints(MaybeCause::Overflow {
-                suggest_increasing_limit: true,
-            }));
+        // during generalization. This prevents hangs caused by an exponential blowup,
+        // see tests/ui/traits/next-solver/coherence-alias-hang.rs.
+        //
+        // We don't do so for `NormalizesTo` goals as we erased the expected term and
+        // bailing with overflow here would prevent us from detecting a type-mismatch,
+        // causing a coherence error in diesel, see #131969. We still bail with verflow
```
Comment (Noratrieb) [inline compiler/rustc_next_trait_solver/src/solve/eval_ctxt/canonical.rs:173]:
```suggestion
        // causing a coherence error in diesel, see #131969. We still bail with overflow
```

Comment (compiler-errors) [reply compiler/rustc_next_trait_solver/src/solve/eval_ctxt/canonical.rs:173]: verflow is actually a type-system specific concept :clueless:

Comment (Noratrieb) [reply compiler/rustc_next_trait_solver/src/solve/eval_ctxt/canonical.rs:173]: ah i see, it describes the way crates of different versions flow through the type system

Comment (lcnr) [reply compiler/rustc_next_trait_solver/src/solve/eval_ctxt/canonical.rs:173]: fixed as part of #132026 :3

**Pull request merged**